node/src/tasks/bsp_upload_file.rs

```diff
@@ -252,6 +252,33 @@ where
             .map_err(|_| anyhow::anyhow!("File key and HasherOutT mismatch!"))?;
         self.file_key_cleanup = Some(file_key_hash);
 
+        // Get the node's public key needed for threshold calculation.
+        let node_public_key = self
+            .storage_hub_handler
+            .blockchain
+            .get_node_public_key()
+            .await;
+
+        // Query runtime for the earliest block where the BSP can volunteer for the file.
+        let earliest_volunteer_block = self
+            .storage_hub_handler
+            .blockchain
+            .query_file_earliest_volunteer_block(
+                node_public_key,
+                H256::from_slice(file_key.as_ref()),
+            )
+            .await
+            .map_err(|e| anyhow!("Failed to query file earliest volunteer block: {:?}", e))?;
+
+        // If earliest_volunteer_block is 0 it means that we can volunteer right away.
+        if earliest_volunteer_block > U256::zero() {
+            // TODO: if the earliest block is too far away, we should drop the task.
```

Comment:
How about adding a timout here? Cause it could be either that the block is too far away, or the chain gets stalled and too much time passes by, even if it could have volunteered in a short number of blocks.